File: daily/Arrays/821.Shortest-distance-to-a-character.go
package daily

func shortestToChar(s string, c byte) []int {
	answer := make([]int, len(s))
	n := len(s)
	prev := -n

	for i := 0; i < n; i++ {
		if s[i] == c {
			prev = i
		}
		answer[i] = i - prev
	}

	prev = 2 * n
	for i := n - 1; i >= 0; i-- {
		if s[i] == c {
			prev = i
		}
		answer[i] = min(answer[i], prev-i)
	}
	return answer
}

File: daily/Arrays/821.Shortest-distance-to-a-character.py
from typing import List


class Solution:
    def shortestToChar(self, s: str, c: str) -> List[int]:
        n = len(s)
        ans = [0] * n
        prev = -n
        for i in range(n):
            if s[i] == c:
                prev = i
            ans[i] = i - prev
        prev = 2 * n
        for i in range(n - 1, -1, -1):
            if s[i] == c:
                prev = i
            ans[i] = min(ans[i], prev - i)
        return ans

